연락처관리 UI [STEP 2] Avery, Logan

devxsby opened the pull request:
> 안녕하세요 (@protocorn93)콘 👨🏻‍🌾 
다시 Avery와 Logan 입니다 😆 지난 step1에 올려주셨던 꿀 떨어지는 리뷰가 저희에게 너무 너무 큰 도움이 됐어요! 감사해요! 🫶🏻 이번 Step2도 요구사항이 그렇게 어렵지는 않았기 때문에 저희는 앞으로 진행하게 될 스탭에 대해 논의하는 시간을 더 가지고 구조를 보충하는 쪽으로 공부했었어요. 이번에도 좋은 리뷰 잘 부탁드립니다~~ 🌽🍯

## 📌 Step 2 구현 사항

- [x] TableView 를 활용하여 연락처 목록을 화면에 표시하기
- [x] 각 행의 cell은 subtitle style을 사용
- [x] 각 행(row)에 표시할 항목
    - 이름, 나이, 연락처

![Simulator Screen Recording - iPhone 14 - 2023-02-02 at 11 33 32](https://user-images.githubusercontent.com/80062632/216216973-8ffbcfcf-9104-43a9-8d0a-67edb01cf6e7.gif)


## 🦄 추가 사항
- 안전하게 배열을 조회하기 위해 `Array+` 익스텐션을 추가했어요.
    - `why?` <br> `Array: Out of range` 런타임 error가 발생되면 프로그램이 바로 죽기 때문에 안전하게 조회를 해야한다고 생각했어요.
      따라서, 우리는 만약 out of range이면 nil을 내뱉는 서브 스크립트 메소드를 만들어주었어요.

    ```swift
    extension Array {
        subscript(safe index: Int) -> Element? {
            return indices ~= index ? self[index] : nil
        }
    }
    ```

## 💡 이번 스텝 수행 중 핵심 경험 

- TableView의 기초구현 및 참고 공식 문서
    - [UITableView](https://developer.apple.com/documentation/uikit/uitableview)
    - [Table views](https://developer.apple.com/documentation/uikit/views_and_controls/table_views)
    - [Filling a table with data](https://developer.apple.com/documentation/uikit/views_and_controls/table_views/filling_a_table_with_data)
    - [Table View Programming Guide for iOS](https://developer.apple.com/library/archive/documentation/UserExperience/Conceptual/TableView_iPhone/TableViewStyles/TableViewCharacteristics.html#//apple_ref/doc/uid/TP40007451-CH3-SW1)

## 🐣 PR Point

- 샘플링을 위하여 DummyData를 만들었는데 Person의 DummyData이기 때문에 Person의 extension에 두었는데 해당 객체의 위치가 바람직한지 어디가 가장 적절하다고 생각하시는지 궁금해요. 샘플링을 위한 데이터들이 혹시나 더 있을 경우엔 아예 따로 샘플링을 위한 데이터들을 모아놓을 파일을 만드는 쪽이 나을까요? 🤔

- 프로토콜을 준수하고 프로토콜 요구사항을 구현할때 콘은 코드를 어떻게 구성하시는지 궁금해요. 듣기로는 extension으로 프로토콜을 준수하고 요구사항을 구현하는 방법도 있는것 같고 동작부를 깔끔하게 나누기 위해 extension에 구현하는 등 방법이 여러가지인 것 같아서요 🧐

sloDevz committed:
feat: UITableView 및 UITableVIewCell 추가

sloDevz committed:
feat: UITableView의 UITableViewDelegate, UITableViewDataSource 프로토콜 준수

sloDevz committed:
feat: UITableView에 Cell 추가 및 기본적인 TableView출력 구현

devxsby committed:
merge: main 브랜치 최산화

devxsby committed:
feat: 접근제한자 추가

sloDevz committed:
feat: DummyData 생성 및 TableView와 연결

devxsby committed:
chore: 파일명 및 변수명 수정

devxsby committed:
merge: main 브랜치 최신화

devxsby committed:
merge: main 브랜치 최신화

sloDevz committed:
feat: ArrayExtension 추가

devxsby committed:
chore: 변수명 및 함수명 수정

devxsby committed:
chore: 타겟 버전 변경

- macOS: 12.0
- iOS: 13.0

devxsby committed:
chore: 접근제한자 수정 및 MARK 주석 추가

devxsby committed:
chore: main 함수 private 접근 제한자 추가

protocorn93 commented on ContactManager/ContactManager/Service/InputEditor.swift:
```diff
@@ -11,7 +11,7 @@ protocol InputEditorProtocol {
     func getUserInput() throws -> String
     func getContactInfo() throws -> UserInputModel
-    func requestValidation(with model: UserInputModel) throws -> Person
+    func requestValidation(with inputtedModel: UserInputModel) throws -> Person
```
개인적으론 `inputtedModel` 이란 네이밍이 어색하네요! 저는 해당 타입의 변수명을 만들기 어려울 땐 그냥 해당 타입의 첫 글자만 소문자로 해주곤해요! (`userInputModel`)

devxsby replied:
말씀해주신 부분에 동의해서 [bf37346](https://github.com/tasty-code/ios-contact-manager-ui/pull/13/commits/bf373461348c1b13ec0509835098a54e6699acbc) 여기 커밋에 반영 했어요 ~! 👍🏻

protocorn93 commented on ContactManager/ContactManager/Service/ProgramManager.swift:
```diff
@@ -53,9 +53,9 @@ final class ProgramManager: ProgramManagerProtocol {
 }
 
-extension ProgramManager {
+private extension ProgramManager {
```
물론 `private extension`을 통해 하위 메서드들에 동일하게 접근 제어자 범위를 적용할 수 있지만, 저는 개인적으론 각 메서드에 모두 `private`을 붙이는걸 선호해요! 메서드가 많아진다면 휴먼 에러가 발생할 수 있기 때문이에요~

devxsby replied:
원래 코드에는 각 메서드에 `private` 키워드를 붙여놨었어요.
그런데 다 `private` 키워드를 붙여놔서 이것 또한 리소스 낭비라고 생각이 들어서 바꿔봤어요.
하지만 다시 생각해보니 각 메서드가 같은 접근제한자를 갖는다는 것이 더 이상하다는 생각이 들어서 [0f197f5](https://github.com/tasty-code/ios-contact-manager-ui/pull/13/commits/0f197f551d8687eadf55506e5b30b5bfc73cc27e)에 반영했습니다...! 
감사합니다 🌞

protocorn93 commented on ContactManager/ContactManagerUI/ContactListViewController.swift:
```diff
@@ -0,0 +1,53 @@
+    
+    // MARK: - Methods
+
+    private func setDelegate() {
+        contactListTableView.delegate = self
+        contactListTableView.dataSource = self
+    }
```
이렇게 메서드로 분리해주는 것 좋은 것 같아요. 사소해보일 수 있지만 저는 이런것들이 가독성을 더 좋게한다고 생각해요! 👍

devxsby replied:
사소한 부분이지만 칭찬해주셔서 힘이나요. 감사합니다~~!! ☺️

protocorn93 commented on ContactManager/ContactManagerUI/ContactListViewController.swift:
```diff
@@ -0,0 +1,53 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cellIdentifier = "ContactTableViewCell"
+        let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath)
```
`Array`의 요소를 안전하게 접근할 수 있는 `extension`을 구현해주신 것처럼 cell identifier 역시 거의 만국 공통 extension 코드가 있어요! 한번 찾아보시고 적용해보시면 좋을 것 같아요!

devxsby replied:
감사합니다!! 😊 그런데 하나 질문이 있어요~~!!!
제가 알고있던 유틸은 VC, TVC, CVC의 classNamed을 String으로 가져올 수 있도록 연산 프로퍼티 설정해주는 메서드였어요!

```swift
extension NSObject {
    
    static var className: String {
        NSStringFromClass(self.classForCoder()).components(separatedBy: ".").last!
    }
    
    var className: String {
        NSStringFromClass(self.classForCoder).components(separatedBy: ".").last!
    }
}
```
그런데 저희가 `ContactTableViewCell`을 커스텀 셀로 만들지 않아서 결국 identifier를 저렇게 하드코딩으로 지정해야 한다고 생각해서 저 방식으로 구현했는데 혹시 제가 잘 못찾고 있는 것 같아서 다른 익스텐션이 있는지 키워드를 알려주실 수 있으실까요? 🥲

[Nicer reuse identifiers with protocols in Swift](https://medium.com/bleeding-edge/nicer-reuse-identifiers-with-protocols-in-swift-97d18de1b2df) 이 아티클도 참고해봤는데 잘 못찾겠어요.. 흑 ㅠ

protocorn93 commented on ContactManager/ContactManagerUI/ContactListViewController.swift:
```diff
@@ -0,0 +1,53 @@
+        
+        guard let cellData = dummyContactList[safe: indexPath.row] else { return UITableViewCell() }
+        
+        cell.textLabel?.text = cellData.name + "(\(cellData.age))"
+        cell.detailTextLabel?.text = cellData.phoneNum
```
셀의 UI 요소에 직접 접근하기보단 메서드를 생성해서 값만 전달해주면 좋을 것 같아요.

protocorn93 replied:
앗! 이건 기본 셀이라 이렇게 접근해야겠네요 😂 이후에 커스텀 셀을 만들게 된다면 그렇게 진행해주세요!

devxsby replied:
넵~! 😆

protocorn93 commented on ContactManager/ContactManagerUI/ContactListViewController.swift:
```diff
@@ -0,0 +1,53 @@
+
+    // MARK: - @IBOutlet Properties
+
+    @IBOutlet weak private var contactListTableView: UITableView!
```
`private` 선언 잘 해주셨네요! 👍

devxsby replied:
감사합니다 !!

protocorn93 reviewed (changes requested):
두 분 잘 구현해주셨어요! 이번에는 크게 난이도가 높았던 부분은 없어보여요! 


## 질문

> 샘플링을 위하여 DummyData를 만들었는데 Person의 DummyData이기 때문에 Person의 extension에 두었는데 해당 객체의 위치가 바람직한지 어디가 가장 적절하다고 생각하시는지 궁금해요. 샘플링을 위한 데이터들이 혹시나 더 있을 경우엔 아예 따로 샘플링을 위한 데이터들을 모아놓을 파일을 만드는 쪽이 나을까요? 🤔

음 사실 보통 운영되는 서비스의 타겟에는 더미 데이터를 넣지 않아요. 보통 더미 데이터는 테스트 타겟에서 단위 테스트를 작성할 때 사용하곤 해요. 

그래도 필요하다면 두 분이 진행해주신 것 처럼 해당 타입의 `static` 변수로 선언해주곤 해요!  

> 프로토콜을 준수하고 프로토콜 요구사항을 구현할때 콘은 코드를 어떻게 구성하시는지 궁금해요. 듣기로는 extension으로 프로토콜을 준수하고 요구사항을 구현하는 방법도 있는것 같고 동작부를 깔끔하게 나누기 위해 extension에 구현하는 등 방법이 여러가지인 것 같아서요 

저는 보통 extension을 역할별로 나눠놓곤해요! 단순 프로토콜 준수 뿐만 아니라요! 저는 보통 아래와 같이 작성하곤해요! 

```swift 
final class SomeViewController: UIViewController { 
  // MARK: Properties 
  ...

  // MARK: Initializing
  ...
}


// MARK: - Actions

extension SomeViewController { 

  @objc 
  private func someActions() { 

  }
}


// MARK: - Layouts 

extension SomeViewController { 

  private func defineLayout() { 

  }
}


// MARK: UITableViewDelegate 

extension UIViewController: UITableViewDelegate {

  func tableView(....) { 

  }
}
```

devxsby committed:
chore: ContactListVC private extension 접근 제한자 변경

- private extension으로 한 번에 적용한던 것을 internal extension에 각 메서드에 private 접근제한자를 직접 추가하도록 변경했어요.

devxsby committed:
chore: ContactListVC extension 정리

devxsby committed:
chore: requestValidation 파라미터명 변경

inputtedModel에서 userInputModel로 변경했어요.
